### PR TITLE
Smallchanges

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

On receiving a machine tag after the first time, check to ensure the machine is alive, or close down the go routine.

## QA steps

1. export JUJU_DEV_FEATURE_FLAGS=instance-mutater
2. juju bootstrap localhost
3. juju deploy ./testcharms/charm-repo/quantal/lxd-profile-alt --to lxd -n 2
2. wait for both units to be active
3. juju upgrade-charm lxd-profile-alt  --path ./testcharms/charm-repo/quantal/lxd-profile-alt
4. verify that both are successful and that their profiles have been updated

### CODE REVIEW

**Comment by hmlanigan** [comment]
Ran race.bash for:

OK: 27 passed
PASS
536 successful runs in 25 min 3 sec

----------------------------------------------------------------------
FAIL: worker_test.go:286: workerEnvironSuite.TestCharmProfilingInfoNotProvisioned

new machine 0
DEBUG received tag "machine-0"
TRACE watching change on mutaterMachine 0
worker_test.go:299:
    c.Assert(err, jc.Satisfies, errors.IsNotProvisioned)
... obtained = nil
... func(T) bool func(error) bool = (func(error) bool)(0x69e740)

OOPS: 26 passed, 1 FAILED
--- FAIL: TestPackage (0.05s)
FAIL

File: worker/instancemutater/mutater.go [hunk 1]
```diff
@@ -128,20 +133,39 @@ func (m mutaterMachine) watchProfileChangesLoop(profileChangeWatcher watcher.Not
...
 	if len(info.CurrentProfiles) == 0 && len(info.ProfileChanges) == 0 {
 		// no changes to be made, return now.
 		return nil
 	}
 
+	if err := m.machineApi.Refresh(); err != nil {
+		return err
+	}
+	if m.machineApi.Life() == params.Dead {
+		// Machine is dead, continue onwards as we can't do anything in this
```
Comment (manadart) [inline worker/instancemutater/mutater.go]: I think the comment at the call site is sufficient to indicate what is happening.
This reads a bit like the behaviour changed at some point, but the comment is vestigial. I'd remove it.

File: worker/instancemutater/worker_test.go [hunk 2]
```diff
@@ -189,72 +185,92 @@ var _ = gc.Suite(&workerEnvironSuite{})
...
-		s.expectFacadeMachineTag,
-		s.notifyMachineAppLXDProfile(1, s.closeDone),
-		s.expectMachineTag,
-		s.expectMachineCharmProfilingInfo(3),
-		s.expectLXDProfileNames,
-		s.expectSetCharmProfiles,
+		s.notifyMachines([][]string{{"0"}}),
+		s.expectFacadeMachineTag(0),
+		s.notifyMachineAppLXDProfile(0, 1),
+		s.expectMachineTag(0),
```
Comment (manadart) [inline worker/instancemutater/worker_test.go]: Looks like `expectFacadeMachineTag` already satisfies this.

File: worker/instancemutater/worker_test.go [hunk 3]
```diff
@@ -189,72 +185,92 @@ var _ = gc.Suite(&workerEnvironSuite{})
...
 func (s *workerEnvironSuite) TestVerifyCurrentProfilesTrue(c *gc.C) {
-	defer s.setup(c).Finish()
+	defer s.setup(c, 1).Finish()
+
+	w := s.workerForScenario(c,
+		s.ignoreLogging(c),
+		s.notifyMachines([][]string{{"0"}}),
+		s.expectFacadeMachineTag(0),
+		s.notifyMachineAppLXDProfile(0, 1),
+		s.expectMachineTag(0),
```
Comment (manadart) [inline worker/instancemutater/worker_test.go]: Same here.

File: worker/instancemutater/worker_test.go [hunk 4]
```diff
@@ -189,72 +185,92 @@ var _ = gc.Suite(&workerEnvironSuite{})
...
-			{"0"},
-		}, s.noopDone),
-		s.expectFacadeMachineTag,
-		s.notifyMachineAppLXDProfile(1, s.closeDone),
-		s.expectMachineTag,
-		s.expectCharmProfilingInfoSimpleNoChange,
+		s.notifyMachines([][]string{{"0"}}),
+		s.expectFacadeMachineTag(0),
+		s.notifyMachineAppLXDProfile(0, 1),
+		s.expectMachineTag(0),
```
Comment (manadart) [inline worker/instancemutater/worker_test.go]: And here.

File: worker/instancemutater/worker_test.go [hunk 5]
```diff
@@ -394,39 +419,98 @@ func (s *workerSuite) expectCharmProfilingInfo(mock *mocks.MockMutaterMachine, r
...
+		mExp := s.machine[machine].EXPECT()
+		mExp.Refresh().Return(nil)
+		mExp.Life().Return(params.Alive)
+		mExp.SetModificationStatus(status.Idle, "", nil).Return(nil)
+	}
 }
 
-func (s *workerSuite) expectModificationStatusApplied() {
-	s.machine.EXPECT().SetModificationStatus(status.Applied, "", nil).Return(nil)
+func (s *workerSuite) expectMachineDead(machine int, group *sync.WaitGroup) func() {
```
Comment (manadart) [inline worker/instancemutater/worker_test.go]: I think this one needs a new name now.

File: worker/instancemutater/worker_test.go (hunk 4 above)
Comment (hmlanigan) [reply worker/instancemutater/worker_test.go]: removed

File: worker/instancemutater/worker_test.go (hunk 2 above)
Comment (hmlanigan) [reply worker/instancemutater/worker_test.go]: removed

File: worker/instancemutater/mutater.go (hunk 1 above)
Comment (hmlanigan) [reply worker/instancemutater/mutater.go]: removed

File: worker/instancemutater/worker_test.go (hunk 3 above)
Comment (hmlanigan) [reply worker/instancemutater/worker_test.go]: removed

File: worker/instancemutater/worker_test.go (hunk 5 above)
Comment (hmlanigan) [reply worker/instancemutater/worker_test.go]: done

**Comment by hmlanigan** [comment]
$$merge$$

**Pull request merged**